### PR TITLE
WIP to get qt 4.6.2 support for Alexander

### PR DESCRIPTION
**Pull request by math07**
Two persons on the list ask us to have qt 4.6.2 from 2009 compatibility.
Shinnok,
The commented code are there parts that are not fixed yet, It highlights what cause problem with 4.6.2.

### CODE REVIEW

File: mainwindow.cpp [hunk 1]
```diff
@@ -465,7 +466,7 @@ bool MainWindow::eventFilter(QObject *watched, QEvent *event)
     QWidget* widged = (QWidget*) watched;
     switch (event->type())
     {
-    case QEvent::StyleAnimationUpdate:
+    case 1://QEvent::StyleAnimationUpdate:
```
Comment (math07) [inline mainwindow.cpp]: I put case 1: just to make it compile. What we have to do is replace QEvent::StyleAnimationUpdate: with something defined old version of Qt that Solar want. I think you'd be more suited than me to correct this as you can reproduce the 30% CPU issue on your laptop.

File: mainwindow.cpp [hunk 2]
```diff
@@ -936,7 +937,7 @@ void MainWindow::fillSettingsWithDefaults()
     QStringList possiblePaths;
     QString john;
     // Find john on system path, which is determined by PATH variable
-    QString johnSystemPath = QStandardPaths::findExecutable("john", QStringList());
+    /*QString johnSystemPath = QStandardPaths::findExecutable("john", QStringList());
```
Comment (math07) [inline mainwindow.cpp]: We can no longer use QStandardPaths, kind of sad ... The oldest equivalent seems to be QDesktopServices but I didn't find findExecutable stuff yet.

**Comment by math07** [comment]
Also, I'll make the workaround for the tooltip of the menu to avoid using setToolTipVisible(true) http://stackoverflow.com/questions/27161122/qtooltip-for-qactions-in-qmenu in the multi session PR.

File: mainwindow.ui [hunk 3]
```diff
@@ -66,9 +66,6 @@
       <property name="horizontalScrollBarPolicy">
        <enum>Qt::ScrollBarAlwaysOff</enum>
       </property>
-      <property name="sizeAdjustPolicy">
```
Comment (math07) [inline mainwindow.ui]: This property don't exist on old qt version and we can't put #define in the .ui file so we get a bug where the tab menu takes like 2-3 times more space than needed but the text of the option stay as small with Mot...XXX . Maybe we should reput the workaround that Aleksey did in 2012 that we removed at the beggining of the summer. When we added adjustToContents property, it didn't actually solved the ... bug that we had with translation english to french and incomplete word with ... for the QLIstWidget

Comment (math07) [inline mainwindow.ui]: So we don't really loose anything feature wise and bug-wise by reintroducing Aleksey's workaround, except this property looks cleaner in the code.

**Comment by math07** [comment]
@shinnok What's your opinion about those changes and the inability to use new functions of qt newer than 2009 ?

File: mainwindow.cpp [hunk 4]
```diff
@@ -166,7 +167,7 @@ MainWindow::MainWindow(QSettings &settings)
     //     warnAboutDefaultPathToJohn();
     Translator& translator = Translator::getInstance();
     m_ui->comboBox_LanguageSelection->insertItems(0, translator.getListOfAvailableLanguages());
-    m_ui->comboBox_LanguageSelection->setCurrentText(translator.getCurrentLanguage());
+    //TODO:m_ui->comboBox_LanguageSelection->setCurrentText(translator.getCurrentLanguage());
```
Comment (math07) [inline mainwindow.cpp]: This one seems to be easy to fix by findText() which return index of the language in the comboBox and setCurrentIndex(int).

File: hashtypechecker.h [hunk 5]
```diff
@@ -5,7 +5,8 @@
 
 #include <QObject>
 #include <QTextStream>
-#include <QtConcurrent/QtConcurrent>
+
+#include <vector>
```
Comment (shinnok) [inline hashtypechecker.h]: Let's use QList instead.

File: johnprocess.cpp [hunk 6]
```diff
@@ -27,17 +27,18 @@ void JohnProcess::setupChildProcess()
...
         /* Send sigterm to all processes of the group
          * created previously with setsid()
          * This is done in Johnny because of a bug in john <= 1.8.0
          * where it didn't forward signals to children and we want to keep
          * compatibility with those versions of John for now */
-        ::kill(-processId(), SIGTERM);
+        ::kill(-pid(), SIGTERM);
     }
 #elif defined Q_OS_WIN
+    Q_PID pi = pid();
```
Comment (shinnok) [inline johnprocess.cpp]: Rename the var to pid.

File: mainwindow.cpp (hunk 4 above)
Comment (shinnok) [inline mainwindow.cpp]: Do it like that with an #ifdef  then.

File: mainwindow.cpp (hunk 1 above)
Comment (shinnok) [inline mainwindow.cpp]: Just wrap around an #ifdef !OSX. I don't expect anyone to attempt building with 4.6.2 on OS X.

File: mainwindow.cpp (hunk 2 above)
Comment (shinnok) [inline mainwindow.cpp]: We need to search manually if < Qt5. Use something similar to this https://github.com/Tarsnap/tarsnap-gui/commit/5d9157163c5284bb8712ad0f32b9bf2ba5bcc2fa#diff-fd0081fe4276fa1c3d9f87bc40a8d1d1L355 .

File: mainwindow.cpp [hunk 7]
```diff
@@ -944,7 +945,7 @@ void MainWindow::fillSettingsWithDefaults()
     // John might be in in the same directory than johnny
     QString johnOtherPaths = QStandardPaths::findExecutable("john", QStringList(QDir::currentPath()));
     if(!johnOtherPaths.isEmpty())
-        possiblePaths << johnOtherPaths;
+        possiblePaths << johnOtherPaths;*/
```
Comment (shinnok) [inline mainwindow.cpp]: You can search in the current directory manually.

File: mainwindow.ui (hunk 3 above)
Comment (shinnok) [inline mainwindow.ui]: I'm still considering replacing that with just a toolbar.

Comment (math07) [inline mainwindow.ui]: Okay, I'll try it and see if it works better

File: mainwindow.ui [hunk 8]
```diff
@@ -1162,10 +1024,13 @@
     <rect>
      <x>0</x>
      <y>0</y>
-     <width>884</width>
+     <width>1012</width>
      <height>27</height>
     </rect>
    </property>
+   <property name="contextMenuPolicy">
+    <enum>Qt::PreventContextMenu</enum>
```
Comment (math07) [inline mainwindow.ui]: Using this on any toolbars fix #33

**Comment by math07** [comment]
As you suggested, I adapted the tab selection from QListWidget to a toolbar in 5202c5c. 
Two positive points:
1) It doesn't take way more room than needed (almost half the screen) if we remove  <property name="sizeAdjustPolicy"> which isn't supported in old version of qt. So we don't need to put back Aleksey's 2012 workaround.
2) It's the first solution that works without "..." when we dynamically translate from a language who has smaller words like english "Password" and french "Mot de passe". The layout expands properly with translation like any other widget of MainWindow.

File: hashtypechecker.h (hunk 5 above)
Comment (math07) [inline hashtypechecker.h]: Done

File: johnprocess.cpp (hunk 6 above)
Comment (math07) [inline johnprocess.cpp]: Done

File: mainwindow.cpp (hunk 4 above)
Comment (math07) [inline mainwindow.cpp]: Done

File: mainwindow.cpp (hunk 1 above)
Comment (math07) [inline mainwindow.cpp]: Done

File: mainwindow.cpp (hunk 2 above)
Comment (math07) [inline mainwindow.cpp]: Done

**Comment by math07** [comment]
This task is done and compile on qt 4.6.2(successfully compiled on centOS 6).
Note that while running johnny(./johnny) we get some warning in the terminal on centOS 6. It doesn't seem to do anything bad to johnny. By googling the warning, it seems to affect most qt 4 applications on the distros like centOs 6.
 (process:2412): Gtk-CRITICAL **: gtk_widget_style_get: assertion
`GTK_IS_WIDGET (widget)' failed

File: mainwindow.cpp [hunk 9]
```diff
@@ -19,11 +19,16 @@
...
 #include <QThread>
 #include <QTextCursor>
-#include <QStandardPaths>
+#include <QDesktopServices>
 #include <QInputDialog>
+#include <QtDebug>
 
 #define INTERVAL_PICK_CRACKED 600
-#define PASSWORD_TAB 0
+#define PASSWORD_TAB    0
```
Comment (shinnok) [inline mainwindow.cpp]: It's usually nicer to prefix with the common word. This way you get nice tab completion and it's a bit more evident what they're for.
TAB_PASSWORDS
TAB_OPTIONS
TAB_STATISTICS
....

File: mainwindow.cpp [hunk 10]
```diff
@@ -19,11 +19,16 @@
 #include <QClipboard>
 #include <QThread>
 #include <QTextCursor>
-#include <QStandardPaths>
```
Comment (shinnok) [inline mainwindow.cpp]: Shouldn't you have kept this in an ifdef? You might get warnings or issues in certain scenarios.

**Comment by math07** [comment]
Changes done. The removal of the listWidget with the toolbar will need conflict resolution with multipleSessionManagement PR. Merge any of the two first and I'll manually fix conflicts in the other for you to be able to automatically merge both pull requests like I always do.

**Comment by math07** [comment]
I resolved merge conflicts in one commit and made some modifications regarding qt 4.6.2 support in a new commit. This is ready to be merged

**Comment by shinnok** [comment]
After you attend my recently added comments I'll merge this.

**Comment by math07** [comment]
I think the coding style with the newlines should be fixed in all source files with my two new commits

**Pull request merged**